### PR TITLE
Move register_translation from skimage.feature to skimage.registration

### PR DESCRIPTION
**Pull request by rfezzani**
## Description

This PR proposes to deprecate `skimage.feature.register_transtlation` in favor of `skimage.registration.register_transtlation`.

## Checklist

- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

## For reviewers

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

### CODE REVIEW

File: skimage/feature/__init__.py [hunk 1]
```diff
@@ -27,6 +28,17 @@
 from .haar import (haar_like_feature, haar_like_feature_coord,
                    draw_haar_like_feature)
 
+from ..registration.register_translation import (
+    register_translation as _register_translation)
```
Comment (jni) [inline skimage/feature/__init__.py]: This makes my head spin. Can you rename the file something like `_fourier_space_translation` or somesuch? In general, we have tried to avoid files (modules) having the exact same name as functions because then it is ambiguous when you write `registration.register_translation` whether you mean the function or the module.

File: skimage/registration/__init__.py [hunk 2]
```diff
@@ -1 +1,8 @@
 from ._optical_flow import optical_flow_tvl1
+from .register_translation import register_translation
```
Comment (jni) [inline skimage/registration/__init__.py]: See above.

File: skimage/registration/__init__.py [hunk 3]
```diff
@@ -1 +1,8 @@
...
+
+
+__all__ = [
+    'optical_flow_tvl1',
+    'register_translation'
+    ]
```
Comment (jni) [inline skimage/registration/__init__.py]: We need to think about the API here. `registration.register_translation` no longer makes sense, it would be kind of like `filters.gaussian_filter` rather than `filters.gaussian`.

Some alternatives: `registration.fourier_cross_correlation`, `registration.translation`, `registration.maximum_cross_correlation`, `registration.phase_shift`...

File: skimage/feature/__init__.py (hunk 1 above)
Comment (jni) [reply skimage/feature/__init__.py]: (Note: I realise that the original file doesn't follow this convention. We haven't been great at enforcing this policy but I think it's worth enforcing going forward.)

File: skimage/registration/__init__.py (hunk 3 above)
Comment (rfezzani) [reply skimage/registration/__init__.py]: I personally vote for `phase_shift`

Comment (jni) [reply skimage/registration/__init__.py]: @scikit-image/core anyone else want to weigh in here?

Comment (grlee77) [reply skimage/registration/__init__.py]: Perhaps `phase_correlation`? The original publication describing it is apparently:
```
Kuglin, C., Hines, D., 1975. The phase correlation image alignment method. In: Proc.Internat. Conf. on Cybernetics and Society, IEEE, pp. 163–165
```
and that is also the title of it's [Wikipedia article](https://en.wikipedia.org/wiki/Phase_correlation).

OpenCV has [phaseCorrelate](https://docs.opencv.org/4.2.0/d7/df3/group__imgproc__motion.html#ga552420a2ace9ef3fb053cd630fdb4952) to calculate the shift.

Comment (sciunto) [reply skimage/registration/__init__.py]: There is also the function match_template that would be good to move and rename as well. For this one, it uses a zero normalized cross correlation (https://en.wikipedia.org/wiki/Cross-correlation#Terminology_in_image_processing)

I suggest to think for these two functions at ones, and I would prefer phase_cross_correlation and zero_normalized_cross_correlation.

Comment (rfezzani) [reply skimage/registration/__init__.py]: @sciunto, even if `template_matching` uses zncc internally, I don't think that the function name should be modified: template matching operation is well identified. Concerning `phase_cross_correlation` suggestion, why not, but if `phase_correlation` is well accepted in the community, why not keep it?

Comment (rfezzani) [reply skimage/registration/__init__.py]: Shouldn't the reference provided by @grlee77 be added to the documentation?

Comment (sciunto) [reply skimage/registration/__init__.py]: template_matching can be used to do other things than matching a template, for instance, you can use it to find shifts between images. Actually, the day I wanted to perform zncc, it took me some time to figure out that template_matching was actually what I wanted to use. That's why I advocate for changing this name.

Comment (rfezzani) [reply skimage/registration/__init__.py]: OK @sciunto, that's an interesting point!
The `*_cross_correlation` functions must then coherent behavior. What if the `*_cross_correlation` functions return a `response` images as `match_template` does and make the `register_translation` return a unique `shift` based on any cross correlation method?
The [`_compute_error`](https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/register_translation.py#L90) function may then be exposed and the [`return_error`](https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/register_translation.py#L134) argument becomes obsolete!

Comment (sciunto) [reply skimage/registration/__init__.py]: That sounds like a good idea. Xref: #3494

Comment (rfezzani) [reply skimage/registration/__init__.py:8]: Do we all agree with `phase_cross_correlation` or do we choose `phase_correlation` instead?

@sciunto, I propose to address #3494 in an other PR ;-)

Comment (sciunto) [reply skimage/registration/__init__.py:8]: yes, sounds good.

**Comment by rfezzani** [comment]
I renamed `register_translation` to `phase_cross_correlation`, according to our previous discussion, any objection?

File: doc/examples/registration/plot_register_translation.py [hunk 4]
```diff
@@ -1,14 +1,13 @@
-"""
-=====================================
+"""=====================================
```
Comment (jni) [inline doc/examples/registration/plot_register_translation.py:1]: revert this change?

File: doc/examples/registration/plot_register_rotation.py [hunk 5]
```diff
@@ -83,7 +84,8 @@
 plt.show()
 
 # setting `upsample_factor` can increase precision
-tparams = register_translation(image_polar, rescaled_polar, upsample_factor=20)
+tparams = phase_cross_correlation(image_polar, rescaled_polar,
+                                  upsample_factor=20)
 shifts, error, phasediff = tparams
```
Comment (jni) [inline doc/examples/registration/plot_register_rotation.py:89]:
```suggestion
shifts, error, phasediff = phase_cross_correlation(
    image_polar, rescaled_polar, upsample_factor=20
)
```

Comment (jni) [reply doc/examples/registration/plot_register_rotation.py:89]: I might also do, `from skimage import registration` ... `registration.phase_cross_correlation(...)`

File: skimage/feature/__init__.py [hunk 6]
```diff
@@ -28,6 +29,16 @@
                    draw_haar_like_feature)
 
 
+@deprecated(alt_func='skimage.registration.register_translation',
+            removed_version='0.19')
+def register_translation(src_image, target_image, upsample_factor=1,
+                         space="real", return_error=True):
+    from ..registration._phase_cross_correlation import (
+        phase_cross_correlation as func)
```
Comment (jni) [inline skimage/feature/__init__.py:37]: Why as func???

Comment (sciunto) [reply skimage/feature/__init__.py:37]: Because func is short, func is universal, func is the solution to any issue, because the secret goal of this library is to provide a unique function func, that will perform a unified image processing.

Comment (rfezzani) [reply skimage/feature/__init__.py:37]: > Because func is short,

That's the main reason! it eases the formatting in the next line.
Since it is a local variable and that this code bloc will be removed in v0.19, I supposed that it is fine...

**Comment by rfezzani** [comment]
:tada: The CI is finally all green!

**Comment by jni** [comment]
@scikit-image/core anyone wanna push the green button? A reminder to squash the commits. =)

**Comment by sciunto** [comment]
I'm on it.

File: skimage/feature/tests/test_masked_register_translation.py [hunk 7]
```diff
@@ -230,14 +232,13 @@ def test_cross_correlate_masked_autocorrelation_trivial_masks():
...
-    arr2 = camera()
 
     # Random masks with 75% of pixels being valid
     m1 = np.random.choice([True, False], arr1.shape, p=[3 / 4, 1 / 4])
-    m2 = np.random.choice([True, False], arr2.shape, p=[3 / 4, 1 / 4])
+    m2 = np.random.choice([True, False], arr1.shape, p=[3 / 4, 1 / 4])
 
-    xcorr = cross_correlate_masked(arr1, arr1, m1, m1, axes=(0, 1),
-                 mode='same', overlap_ratio=0).real
+    xcorr = cross_correlate_masked(arr1, arr1, m1, m2, axes=(0, 1),
```
Comment (sciunto) [inline skimage/feature/tests/test_masked_register_translation.py:240]: Whaoo, that was a bug.

File: skimage/registration/_phase_cross_correlation.py [hunk 8]
```diff
@@ -118,31 +117,33 @@ def register_translation(src_image, target_image, upsample_factor=1,
...
         Reference image.
-    target_image : array
-        Image to register.  Must be same dimensionality as ``src_image``.
+    moving_image : array
+        Image to register. Must be same dimensionality as
+        ``reference_image``.
     upsample_factor : int, optional
         Upsampling factor. Images will be registered to within
-        ``1 / upsample_factor`` of a pixel. For example
+        ``1 /upsample_factor`` of a pixel. For example
```
Comment (sciunto) [inline skimage/registration/_phase_cross_correlation.py:127]: Is this change intended?

Comment (sciunto) [inline skimage/registration/_phase_cross_correlation.py:127]:
```suggestion
        ``1 / upsample_factor`` of a pixel. For example
```

Comment (rfezzani) [reply skimage/registration/_phase_cross_correlation.py]: Nop, sorry :pray:

**Comment by sciunto** [comment]
Perfect! Thank you @rfezzani
